[FIX] account: make fixed and percent included taxes independent

stpa-odoo opened the pull request:
### Steps to reproduce the issue:

1. Create two Taxes, one with Fixed "Tax Computation" and one with Percentage of Price.
2. Set "Included in Price" to True and and "Affect Base of Subsequent Taxes" to False for both.
3. Create an Invoice with one line, add both Taxes to it, make sure each Tax has a different amount.
4. Change the order of the Taxes in the Taxes menu
5. Refresh the Invoice Line (e.g.: remove and put back one of the taxes)
6. The amount of the "percent" Tax has changed along with the Untaxed Amount of the Invoice

### Explanation:

In `_prepare_taxes_computation`, taxes are preprocessed with a priority set on fixed taxes, then taxes with `price_include` set to True, both in descending order and lastly, taxes with `price_include` set to False in ascending order. In `_propagate_extra_taxes_base`, the currently preprocessed tax is also added to the base of other taxes depending on the specifications of those taxes.
When `price_include` is set to True, taxes are never being added to the base of the ones with a higher sequence, it should only be the case if `include_base_amount` is set to True as well.

### Fix reasoning:

The computation should be consistent regardless of the order of the taxes, as they are not supposed to have an influence on each other. The special modes computation are changed accordingly.
The change to the `total_included` special mode for non `_original_price_include` batches fixes an inconsistency when all taxes are excluded, added a test for it.

opw-4068781

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/176150.png)](https://mergebot.odoo.com/odoo/odoo/pull/176150)

nle-odoo reviewed (approved):
seems better but risky (because we could change the price all of a sudden)

but functionally I don't know

stpa-odoo commented:
> seems better but risky (because we could change the price all of a sudden)

The issue reported is an inconsistency between `17.0` and `saas-17.2`, the latter being the faulty version. Now if it is too risky, we can restrict that change to master only.

> but functionally I don't know

I have discussed the case with LAS before opening the PR. It is correct on the functional side, but could see some improvements on the techincal side.

yosuanicolaus commented on addons/account/models/account_tax.py:
```diff
@@ -795,27 +799,9 @@ def add_extra_base(other_batch, tax_data, sign):
                 # To compute the base amount of t2, we need to add back the tax amount of t1.
                 elif special_mode == 'total_excluded':
```
Since we now handle all 3 `special_mode`, maybe it's better to write `else` with comment here just for more clarity. (same for the JS part too, but ignore it if you disagree)
```else:  # special_mode == 'total_excluded'```

yosuanicolaus commented on addons/account/models/account_tax.py:
```diff
@@ -783,7 +783,11 @@ def add_extra_base(other_batch, tax_data, sign):
                 # However, since t1 is not price-included, its base amount is computed by removing first the tax amount of t2.
-                if not special_mode:
+                if not special_mode or special_mode == 'total_included':
+                    if not batch['include_base_amount']:
+                        for other_batch in batches_after:
+                            if other_batch['_original_price_include']:
+                                add_extra_base(other_batch, tax_data, -1)
```
We should also add some comments here why we're removing some of the tax amount from `batches_after` here

yosuanicolaus reviewed (changes requested):
Well done! I just have some small comments, otherwise it's good for me

yosuanicolaus reviewed (approved):
LGTM

final review by @gawa-odoo

stpa-odoo commented:
Hello, I added an example to each section that I changed, they are mostly about the same kind of use case, since the goal of every change is the same.

> For example, I don't really understand this change https://github.com/odoo/odoo/pull/176150/files#diff-1ee3fce434db0c4e897973eebf5c1be501196cbd413e6c250ecc973238f939b9R802

`price_include` depends on `special_mode`, its value is always `False` in the current case, `_original_price_include` is independent. To reproduce the correct behaviour, this change is mandatory.

Regarding this comment and section that I removed: https://github.com/odoo/odoo/blob/8690e98b058923ac5784210f4af61ed8ae1e5e02/addons/account/models/account_tax.py#L806-L822 For the price-include part, it is the same as the normal behaviour section and comment of a price-include tax: https://github.com/odoo/odoo/blob/8690e98b058923ac5784210f4af61ed8ae1e5e02/addons/account/models/account_tax.py#L785-L793 The fact that I could merge both with my fix as well supports this observation.

For the price-exclude part of the example, it is out of place since the behaviour is for price-include taxes only. I can maybe add a simpler example in the price-exclude section.

gawa-odoo commented:
I will give it to las, to be sure, as I may probably miss logic.
But your new examples helped me understand your changes, thank you.
It seems coherent to me (and your changes in the test too).

smetl reviewed (approved):
@robodoo r+
